core/server/common/src/main/java/alluxio/master/journal/JournalTool.java

```diff
@@ -239,4 +212,467 @@ private static void usage() {
+
+      String snapshotDumpFile =
+          String.format("copycat-%s-%s.log", currentSnapshot.index(), currentSnapshot.timestamp());
+
+      try (CheckpointInputStream checkpointStream =
+          new CheckpointInputStream(new SnapshotReaderStream(currentSnapshot.reader()))) {
+        readCheckpoint(checkpointStream, Paths.get(mCheckpointsDir, snapshotDumpFile));
+      } catch (Exception e) {
+        LOG.error("Failed to read from journal.", e);
+      }
+    }
+
+    private NetworkAddressUtils.ServiceType getRaftServiceType() {
+      if (mMaster.contains("job")) {
+        return NetworkAddressUtils.ServiceType.JOB_MASTER_RAFT;
+      } else {
+        return NetworkAddressUtils.ServiceType.MASTER_RAFT;
+      }
+    }
+
+    /**
+     * Writes given entry after goin through range and validity checks.
+     *
+     * @param out out stream to write the entry to
+     * @param entry the entry to write to
+     */
+    private void writeSelected(PrintStream out, JournalEntry entry) {
+      if (entry == null) {
+        return;
+      }
+      Preconditions.checkState(
+          entry.getAllFields().size() <= 1
+              || (entry.getAllFields().size() == 2 && entry.hasSequenceNumber()),
+          "Raft journal entries should never set multiple fields in addition to sequence "
+              + "number, but found %s",
+          entry);
+      if (entry.getJournalEntriesCount() > 0) {
+        // This entry aggregates multiple entries.
+        for (JournalEntry e : entry.getJournalEntriesList()) {
+          writeSelected(out, e);
+        }
+      } else if (entry.toBuilder().clearSequenceNumber().build()
+          .equals(JournalEntry.getDefaultInstance())) {
+        // Ignore empty entries, they are created during snapshotting.
+      } else {
+        if (isSelected(entry)) {
+          out.println(entry);
+        }
+      }
+    }
+
+    /**
+     * Whether an entry is within the range of provided parameters to the tool.
+     *
+     * @param entry the journal entry
+     * @return {@code true} if the entry should be included in the journal log
+     */
+    private boolean isSelected(JournalEntry entry) {
+      long sn = entry.getSequenceNumber();
+      if (sn >= mStart && sn < mEnd) {
+        try {
+          return JournalEntryAssociation.getMasterForEntry(entry).equalsIgnoreCase(mMaster);
+        } catch (IllegalStateException e) {
+          return false;
+        }
+      }
+      return false;
+    }
+
+    /**
+     * Implementation of {@link StateMachine} that echoes its state for dumping.
+     *
+     * EchoJournalStateMachine is attached to the cluster as PASSIVE, which makes it receive all
+     * snapshots happening in the FOLLOWER node that it attaches to. For practical purposes, we
+     * will dump the first snapshot only and stop dumping snapshots afterwards.
+     */
+    public static class EchoJournalStateMachine extends StateMachine implements Snapshottable {
```

Comment:
The expected behavior of copycat passive state is that the state machine provides one snapshot and the journal entries after the snapshot. 

But in our case, we will receive multiple snapshots, then the journal entries that the copycat provides is starting from the first snapshot or from the last snapshot?

Reply:
We'll still get latest snapshot, then entries (if snapshot is not complete). 
However, we'll keep getting snapshots after then. Basically, the passive node follows follower node's internal state, so as the follower takes a snapshot, it's sent to passive node.

Reply:
Got it, thanks. Then by getting the first snapshot + entries, we are able to get the full state.